client/db/bolt/db.go

```diff
@@ -408,6 +414,125 @@ func (db *BoltDB) SetPrimaryCredentials(creds *dexdb.PrimaryCredentials) error {
 	})
 }
 
+func (db *BoltDB) pruneArchivedOrders(prunedSize uint64) error {
+
+	return db.Update(func(tx *bbolt.Tx) error {
+		archivedOB := tx.Bucket(archivedOrdersBucket)
+		if archivedOB == nil {
+			return fmt.Errorf("failed to open %s bucket", string(archivedOrdersBucket))
+		}
+
+		// We won't delete any orders with active matches.
+		activeMatches := tx.Bucket(activeMatchesBucket)
+		if activeMatches == nil {
+			return fmt.Errorf("failed to open %s bucket", string(activeMatchesBucket))
+		}
+		oidsWithActiveMatches := make(map[order.OrderID]struct{}, 0)
+		if err := activeMatches.ForEach(func(k, _ []byte) error {
+			mBkt := activeMatches.Bucket(k)
+			if mBkt == nil {
+				return fmt.Errorf("error getting match bucket %x", k)
+			}
+			var oid order.OrderID
+			copy(oid[:], mBkt.Get(orderIDKey))
+			oidsWithActiveMatches[oid] = struct{}{}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("error building active match order ID index: %w", err)
+		}
+
+		nOrds := uint64(archivedOB.Stats().BucketN - 1 /* BucketN includes top bucket */)
+		if nOrds <= prunedSize {
+			return nil
+		}
+
+		toClear := int(nOrds - prunedSize)
+
+		type orderStamp struct {
+			oid   []byte
+			stamp uint64
+		}
+		deletes := make([]*orderStamp, 0, toClear)
+		sortDeletes := func() {
+			sort.Slice(deletes, func(i, j int) bool {
+				return deletes[i].stamp < deletes[j].stamp
+			})
+		}
+		if err := archivedOB.ForEach(func(oidB, v []byte) error {
+			var oid order.OrderID
+			copy(oid[:], oidB)
+			if _, found := oidsWithActiveMatches[oid]; found {
+				return nil
+			}
+			oBkt := archivedOB.Bucket(oidB)
+			if oBkt == nil {
+				return fmt.Errorf("no order bucket iterated order %x", oidB)
+			}
+			stampB := oBkt.Get(updateTimeKey)
+			if stampB == nil {
+				// Highly improbable.
+				stampB = make([]byte, 8)
+			}
+			stamp := intCoder.Uint64(stampB)
+			if len(deletes) < toClear {
+				deletes = append(deletes, &orderStamp{
+					stamp: stamp,
+					oid:   oidB,
+				})
+				sortDeletes()
+				return nil
+			}
```

Comment:
Could avoid sorting until  `len(deletes) == toClear`:

```suggestion
			if len(deletes) < toClear {
				deletes = append(deletes, &orderStamp{
					stamp: stamp,
					oid:   oidB,
				})
				return nil
			}
			if !sorted {
			    sortDeletes()
			    sorted = true
			}
```